text/0000-const-type-inference.md

```diff
@@ -0,0 +1,114 @@
+- Feature Name: const_type_inference
+- Start Date: 2023-12-21
+- RFC PR: [rust-lang/rfcs#3546](https://github.com/rust-lang/rfcs/pull/3546)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow type inference for `const` or `static` when the type of the initial value is known.
+
+# Motivation
+[motivation]: #motivation
+
+Rust currently requires explicit type annotations for `const` and `static` items.
+
+
+In simple cases, explicitly writing out
+the type of the const seems trivial. However, this isn't always the case:
+
+- Sometimes the constant's value is complex, making the explicit type overly verbose.
+- In some cases, the type may be unnameable.
+- When creating macros, the precise type might not be known to the macro author.
+- Code generators may not have enough information to easily determine the type.
+
+This change aims to make Rust code more concise and maintainable, especially in scenarios where the types of
+const items are complicated or not easily expressible.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You may declare constants and static variables without specifying their types when the type can be inferred
+from the initial value. For example:
+
+```rs
+const PI = 3.1415; // inferred as f64
+static MESSAGE = "Hello, World!"; // inferred as &'static str
+const FN_PTR = std::string::String::default; // inferred as fn() -> String
```

Comment:
technically the `const` would have the function item's type instead of a function pointer type.
https://doc.rust-lang.org/reference/types/function-item.html

Reply:
This is actually a good point - I guess there's not much point to coerce the function item's type into function pointer type for `const` items. But should we coerce for `static` items?

Reply:
I'm confused about why a person would want it as a static at all, so perhaps we shouldn't allow it at all in the first version of this.

Reply:
@Lokathor maybe they just use the const as a quick switch between two cfg implementations?

Reply:
It makes sense as a `const` but little sense as a `static`

Reply:
when the type is a ZST, like function item types, at runtime, `const` and `static` are essentially interchangable. when the type is a function pointer, unless wrapped in something with interior mutability, `static` is basically just a `const` with a stable address, IIRC LLVM will still constant-propagate the value to most uses, since it's marked read-only so LLVM knows the value won't change.

Reply:
Doing implicit coercions into an unknown target type sounds very confusing. If the right-hand side is `String::default`, then the resulting type should be the actual type of the expression, the function item ZST.